### PR TITLE
Step 11

### PR DESCRIPTION
**Pull request by legendado**

### CODE REVIEW

File: app/queries/orders_query.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  FILTERS = %w[in_process delivered in_delivery canceled].freeze
+  DEFAULT_FILTER = :in_process
+
+  def initialize(params:, user:)
+    @params = params
+    @user = user
+  end
+
+  def call
+    return Order.where(user: @user).where('status > ?', 1).order(created_at: :desc) unless @params[:filter]
```
Comment (eugene1599) [inline app/queries/orders_query.rb]: status > 1?
try to refactor

File: spec/controllers/orders_controller_spec.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
+RSpec.describe OrdersController, type: :controller do
+  let(:user) { create(:user) }
+  let(:order) { create(:order, :with_delivery, :with_credit_card, :with_order_items, user: user, status: 3) }
```
Comment (eugene1599) [inline spec/controllers/orders_controller_spec.rb]: 3? use symbol

File: db/seeds.rb [hunk 3]
```diff
@@ -1,8 +1,19 @@
...
 
-Category.all.each do |category|
+Category.find_each do |category|
   FactoryBot.create_list(:book, rand(1..5), :with_images, category: category)
 end
 
-FactoryBot.create_list(:coupon, 3)
-FactoryBot.create_list(:delivery, 3)
+Book.find_each do |book|
+  FactoryBot.create(:review, status: 1, book: book, user: FactoryBot.create(:user))
```
Comment (eugene1599) [inline db/seeds.rb]: 1?

File: spec/controllers/orders_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  describe '#index' do
+    before { get :index }
+
+    it { is_expected.to render_template(:index) }
+    it { is_expected.to respond_with(:ok) }
+  end
+
+  describe '#show' do
+    before do
+      create(:book)
```
Comment (eugene1599) [inline spec/controllers/orders_controller_spec.rb]: why we need this book?

File: spec/factories/orders.rb [hunk 5]
```diff
@@ -1,11 +1,37 @@
 FactoryBot.define do
   factory :order do
     token { SecureRandom.hex(16) }
+    status { rand(0..5) }
+
+    after(:create) do |order|
+      if order.read_attribute_before_type_cast(:status) > 1
```
Comment (eugene1599) [inline spec/factories/orders.rb]: try to refactor this line

File: spec/features/orders/index_spec.rb [hunk 6]
```diff
@@ -0,0 +1,14 @@
+RSpec.describe 'Orders page', type: :feature do
+  let(:user) { create(:user) }
+  let!(:order) { create(:order, status: 3, user: user).decorate }
```
Comment (eugene1599) [inline spec/features/orders/index_spec.rb]: use symbol

File: spec/queries/orders_query_spec.rb [hunk 7]
```diff
@@ -0,0 +1,33 @@
+RSpec.describe OrdersQuery do
+  subject(:result) { described_class.call(params: params, user: user) }
+
+  let(:user) { create(:user) }
+  let!(:in_process_order) { create(:order, status: 2, user: user) }
```
Comment (eugene1599) [inline spec/queries/orders_query_spec.rb]: change statuses to symbol

File: spec/rails_helper.rb [hunk 8]
```diff
@@ -11,7 +11,7 @@
 require 'capybara/rails'
 require 'devise'
 
-Dir[Rails.root.join('spec', 'support', '**', '*.rb')].each { |f| require f }
+Dir[Rails.root.join('spec/support/**/*.rb')].each { |f| require f }
```
Comment (eugene1599) [inline spec/rails_helper.rb]: ???
do you use rubocop?

File: app/assets/javascripts/orders.js [hunk 9]
```diff
@@ -0,0 +1,9 @@
+$(document).on('turbolinks:load', function() {
+  $("tr[data-href]").click(function() {
```
Comment (eugene1599) [inline app/assets/javascripts/orders.js]: try not to use universal selector

File: app/assets/javascripts/orders.js [hunk 10]
```diff
@@ -0,0 +1,9 @@
...
+    window.location = $(this).data("href")
+  });
+
+  $("table[data-href]").click(function() {
```
Comment (eugene1599) [inline app/assets/javascripts/orders.js]: try not to use universal selector

File: app/assets/stylesheets/bookstore/_general.scss [hunk 11]
```diff
@@ -16,6 +16,9 @@ img {
     }
   }
 }
+*[data-href] {
```
Comment (eugene1599) [inline app/assets/stylesheets/bookstore/_general.scss]: try not to use universal selector

File: Gemfile [hunk 12]
```diff
@@ -3,45 +3,31 @@ git_source(:github) { |repo| "https://github.com/#{repo}.git" }
...
+gem 'bootsnap', '>= 1.1.0', require: false
+gem 'country_select', '~> 4.0'
+gem 'devise', '~> 4.7', '>= 4.7.1'
+gem 'draper', '~> 3.1'
+gem 'factory_bot_rails', '~> 5.0', '>= 5.0.2'
+gem 'ffaker', '~> 2.12'
+gem 'haml', '~> 5.1', '>= 5.1.2'
+gem 'haml-rails', '~> 2.0', '>= 2.0.1'
+gem 'image_processing', '~> 1.9', '>= 1.9.3'
+gem 'jbuilder', '~> 2.5'
```
Comment (eugene1599) [inline Gemfile]: do you need this gem?

File: spec/rails_helper.rb (hunk 8 above)
Comment (legendado) [reply spec/rails_helper.rb]: yes

File: app/queries/orders_query.rb [hunk 13]
```diff
@@ -0,0 +1,21 @@
+class OrdersQuery < ApplicationQuery
+  FILTERS = %w[in_process delivered in_delivery canceled].freeze
```
Comment (eugene1599) [inline app/queries/orders_query.rb]: rename constant

File: app/queries/orders_query.rb [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+  DEFAULT_FILTER = :in_process
+
+  def initialize(params:, user:)
+    @params = params
+    @user = user
+  end
+
+  def call
+    return Order.where(user: @user).where(status: FILTERS).order(created_at: :desc) unless @params[:filter]
```
Comment (eugene1599) [inline app/queries/orders_query.rb]: refactor this method

File: spec/controllers/reviews_controller_spec.rb [hunk 15]
```diff
@@ -6,16 +6,7 @@
...
-      let(:params) do
-        { review: {
-          user_id: user.id,
-          book_id: book.id,
-          title: review.title,
-          body: review.body,
-          rating: review.rating
-        } }
-      end
+      let(:params) { { review: attributes_for(:review).merge(book_id: book.id) } }
```
Comment (eugene1599) [inline spec/controllers/reviews_controller_spec.rb]: you can write this without merge


`book: book`

File: spec/factories/orders.rb [hunk 16]
```diff
@@ -1,11 +1,39 @@
 FactoryBot.define do
+  STATUSES = %w[in_process in_delivery delivered canceled].freeze
```
Comment (eugene1599) [inline spec/factories/orders.rb]: dry (you already have this constant)

File: app/models/order.rb [hunk 17]
```diff
@@ -21,6 +19,10 @@ def total
     decorate.sub_total - decorate.discount_price + delivery_price
   end
 
+  def self.completed_statuses
+    statuses.keys[2, 4]
```
Comment (eugene1599) [inline app/models/order.rb]: COMPLETED_STATUSED = %w[status1 status2 status3]

File: app/queries/orders_query.rb [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+  end
+
+  def call
+    Order.where(user: @user).where(status: filter_param).order(created_at: :desc)
+  end
+
+  private
+
+  def filter_param
+    return Order.completed_statuses unless @params[:filter]
```
Comment (eugene1599) [inline app/queries/orders_query.rb]: @params[:filter].presence_in(Order::COMPLETED_STATUSES) || DEFAULT_FILTER

File: app/queries/orders_query.rb [hunk 19]
```diff
@@ -0,0 +1,20 @@
...
+  end
+
+  def call
+    Order.where(user: @user).where(status: filter_param).order(created_at: :desc)
+  end
+
+  private
+
+  def filter_param
+    return Order::COMPLETED_STATUSES unless @params[:filter]
```
Comment (eugene1599) [inline app/queries/orders_query.rb]: delete default filter

and also add ability to add book to cart few times

**Pull request merged**